roles/pcp/tasks/pmie.yml

```diff
@@ -60,5 +60,5 @@
 - name: Ensure performance metric inference is running and enabled on boot
   service:
     name: pmie
-    state: started
+    state: restarted
```

Comment:
Note that this makes the role non-idempotent since it will report `changed` every time.  https://docs.ansible.com/ansible/latest/collections/ansible/builtin/service_module.html#parameter-state

Reply:
Indeed... I tried these handlers instead, but the handler is invoked after the test failed. So, it does not help... :p
```
diff --git a/roles/pcp/tasks/pmie.yml b/roles/pcp/tasks/pmie.yml
index 106dfd7..602eea8 100644
--- a/roles/pcp/tasks/pmie.yml
+++ b/roles/pcp/tasks/pmie.yml
@@ -9,6 +9,7 @@
     group: root
     mode: 0755
   loop: "{{ __pcp_pmieconf_groups|default([]) }}"
+  notify: restart pmie
 
 - name: Ensure extra performance rule group link directories exist
   file:
@@ -18,6 +19,7 @@
     group: root
     mode: 0755
   loop: "{{ __pcp_pmieconf_groups|default([]) }}"
+  notify: restart pmie
 
 - name: Ensure extra performance rules are installed for targeted hosts
   copy:
@@ -27,6 +29,7 @@
     group: root
     mode: '0644'
   loop: "{{ __pcp_pmieconf_rules|default([]) }}"
+  notify: restart pmie
 
 - name: Ensure extra rules symlinks have been created for targeted hosts
   file:
@@ -35,6 +38,7 @@
     state: link
     force: yes
   loop: "{{ __pcp_pmieconf_rules|default([]) }}"
+  notify: restart pmie
 
 - name: Enable performance metric inference for targeted hosts (with control.d)
   template:
@@ -60,5 +64,5 @@
 - name: Ensure performance metric inference is running and enabled on boot
   service:
     name: pmie
-    state: restarted
+    state: started
     enabled: yes
```

Reply:
I guess we could do the similar thing on pmie.yml, too...?
https://github.com/performancecopilot/ansible-pcp/blob/main/roles/pcp/tasks/pmcd.yml#L91-L100

Reply:
Updated the patch based on this idea. All the metrics tests passed.

Reply:
> I guess we could do the similar thing on pmie.yml, too...? https://github.com/performancecopilot/ansible-pcp/blob/main/roles/pcp/tasks/pmcd.yml#L91-L100

Updated the commit based on this idea. With this change, all the metrics tests passed.

Reply:
> We have to do this in a few other roles, where we have to restart the service immediately rather than doing it in the handler.

Ok. Let me take a look. Any particular roles in your mind?

>  You'll probably have to add a suppression for ansible-lint or ansible-test so that it won't tell you to use a handler.
    
Sorry for my ignorance. I'm trying to get the errors/warnings without adding any tags to skip_list in .ansible-lint. The command line `ansible-lint roles/pcp/tasks/pm*.yml` returns these two warnings. Does that mean the direct calling `service restart` is not warned? Or do I have to give some extra options?
```
warn_list:  # or 'skip_list' to silence them completely
  - fqcn-builtins  # Use FQCN for builtin actions.
  - var-spacing  # Jinja2 variables and filters should have spaces before and after.
```

Reply:
for example https://github.com/linux-system-roles/kernel_settings/blob/master/tasks/main.yml#L95 - the task has a `# noqa no-handler` comment

Reply:
Ok. If I remove the `# noqa no-handler` comment from the kernel_settings file, I indeed get the `no-handler` warning.
```
$ ansible-lint tasks/main.yml
warn_list:  # or 'skip_list' to silence them completely
  - fqcn-builtins  # Use FQCN for builtin actions.
  - no-handler  # Tasks that run when changed should likely be handlers.
```
But with no such comment, the `no-handler` is not reported if I run `ansible-lint` with 'roles/pcp/tasks/pm*.yml`... That makes me wonder. I was hoping to see the warning without `# noqa no-handler` and the comment suppresses it...

Reply:
Ok - I wasn't sure if ansible-lint would flag the usage - if ansible-lint is ok with your code, then you don't have to do anything - I just didn't want to put a blanket `- no-handler` in `.ansible-lint` that would affect all role code

Reply:
In general, using `ignore_errors: true` for any task is considered bad form in Ansible - the Ansible way is to use `register` and `failed_when` if the task can return deliberate errors that should be handled.

However, in the case of https://github.com/performancecopilot/ansible-pcp/blob/main/roles/elasticsearch/tasks/main.yml#L63 - we determined that there was no good way of "catching" the error here and doing something meaningful, so we decided to use `ignore_errors: true` here.  Same with https://github.com/performancecopilot/ansible-pcp/blob/main/roles/spark/tasks/main.yml#L69 (I guess - I don't remember that particular problem - hopefully it uses `ignore_errors: true` deliberately and not because the code was copy/pasted from the elasticsearch code).

And - I guess the ansible pcp authors decided to use `ignore-errors` in `.ansible-lint` rather than using a `# noqa ignore-errors` in-line in the task - using the comment in-line in the code is preferred because
* it makes it explicit that the choice to use `ignore_errors: true` is deliberate and intentional
* it doesn't cover up other incorrect usage of `ignore_errors: true` - it forces the developer to evaluate whether or not the use of  `ignore_errors: true` is justified

Reply:
> [...]
>     * it makes it explicit that the choice to use `ignore_errors: true` is deliberate and intentional
> 
>     * it doesn't cover up other incorrect usage of `ignore_errors: true` - it forces the developer to evaluate whether or not the use of  `ignore_errors: true` is justified

The "#noqa ignore-errors" approach wasn't known - it sounds like the role(s) should definitely be updated to use that instead.